File: internal/status/status.go
package status

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/charmbracelet/lipgloss"
	"github.com/mattn/go-isatty"

	"github.com/iamkoch/go-ralph-go/internal/prd"
)

var (
	titleStyle    = lipgloss.NewStyle().Bold(true)
	subtitleStyle = lipgloss.NewStyle().Faint(true)
	passStyle     = lipgloss.NewStyle().Foreground(lipgloss.Color("10"))
	failStyle     = lipgloss.NewStyle().Foreground(lipgloss.Color("9"))
	dimStyle      = lipgloss.NewStyle().Faint(true)
)

// Render returns a styled status string for the given PRD.
func Render(p *prd.PRD, reviewDefault int) string {
	var b strings.Builder

	b.WriteString(titleStyle.Render(p.Project))
	b.WriteString("\n")
	b.WriteString(subtitleStyle.Render(fmt.Sprintf("%s — %d stories", p.BranchName, len(p.UserStories))))
	b.WriteString("\n\n")

	maxIDLen := 0
	maxTitleLen := 0
	for _, s := range p.UserStories {
		if len(s.ID) > maxIDLen {
			maxIDLen = len(s.ID)
		}
		if len(s.Title) > maxTitleLen {
			maxTitleLen = len(s.Title)
		}
	}

	completed := 0
	for _, s := range p.UserStories {
		var icon string
		if s.Passes {
			icon = passStyle.Render("✓")
			completed++
		} else {
			icon = failStyle.Render("✗")
		}

		id := fmt.Sprintf("%-*s", maxIDLen, s.ID)
		title := fmt.Sprintf("%-*s", maxTitleLen, s.Title)
		priority := dimStyle.Render(fmt.Sprintf("P%d", s.Priority))

		b.WriteString(fmt.Sprintf("  %s  %s  %s  %s\n", icon, id, title, priority))
	}

	b.WriteString("\n")
	b.WriteString(fmt.Sprintf("Progress: %d/%d complete\n", completed, len(p.UserStories)))

	next := p.NextIncompleteStory()
	if next == nil {
		next = p.NextReviewStory(reviewDefault)
	}
	if next != nil {
		if next.Passes {
			eff := next.EffectiveReviewPasses(reviewDefault)
			b.WriteString(fmt.Sprintf("Next: %s — %s (review %d/%d)\n", next.ID, next.Title, next.ReviewsCompleted+1, eff))
		} else {
			b.WriteString(fmt.Sprintf("Next: %s — %s\n", next.ID, next.Title))
		}
	}

	return b.String()
}

// Run reads the PRD and prints the status to stdout.
func Run(baseDir string, reviewDefault int) {
	p, err := prd.Read(filepath.Join(baseDir, "prd.json"))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading prd.json: %v\n", err)
		os.Exit(1)
	}
	fmt.Print(Render(p, reviewDefault))
}

// Confirm shows the status and run config, then asks the user to proceed.
// Returns true if the user confirms.
func Confirm(p *prd.PRD, reviewDefault int, tool string, maxIter int) bool {
	fmt.Print(Render(p, reviewDefault))
	fmt.Println()
	fmt.Printf("Tool: %s | Max iterations: %d\n\n", tool, maxIter)

	if !isatty.IsTerminal(os.Stdin.Fd()) {
		return true
	}

	fmt.Print("Press Enter to start, or q + Enter to quit: ")
	reader := bufio.NewReader(os.Stdin)
	line, _ := reader.ReadString('\n')
	line = strings.TrimSpace(line)
	return line != "q" && line != "Q"
}

File: src/commands/status.ts
import {
	type PRD,
	type Story,
	effectiveReviewPasses,
	nextIncompleteStory,
	nextReviewStory,
	readPrd,
} from "../core/prd.js";
import { dim, fail, pass, subtitle, title } from "../tui/styles.js";

export function render(prd: PRD, reviewDefault: number): string {
	const lines: string[] = [];

	lines.push(title(prd.project));
	lines.push(subtitle(`${prd.branchName} — ${prd.userStories.length} stories`));
	lines.push("");

	const maxIdLen = Math.max(...prd.userStories.map((s) => s.id.length));
	const maxTitleLen = Math.max(...prd.userStories.map((s) => s.title.length));

	let completed = 0;
	for (const story of prd.userStories) {
		const icon = story.passes ? pass("\u2713") : fail("\u2717");
		if (story.passes) completed++;

		const id = story.id.padEnd(maxIdLen);
		const storyTitle = story.title.padEnd(maxTitleLen);
		const priority = dim(`P${story.priority}`);

		lines.push(`  ${icon}  ${id}  ${storyTitle}  ${priority}`);
	}

	lines.push("");
	lines.push(`Progress: ${completed}/${prd.userStories.length} complete`);

	let next: Story | null = nextIncompleteStory(prd);
	if (!next) {
		next = nextReviewStory(prd, reviewDefault);
	}
	if (next) {
		if (next.passes) {
			const eff = effectiveReviewPasses(next, reviewDefault);
			lines.push(
				`Next: ${next.id} — ${next.title} (review ${(next.reviewsCompleted ?? 0) + 1}/${eff})`,
			);
		} else {
			lines.push(`Next: ${next.id} — ${next.title}`);
		}
	}

	return lines.join("\n");
}

/** Compact story list for display during run loop. */
export function renderCompact(prd: PRD, maxTitleLen = 35): string {
	const lines: string[] = [];
	for (const story of prd.userStories) {
		const icon = story.passes ? pass("\u2713") : fail("\u2717");
		const storyTitle =
			story.title.length > maxTitleLen
				? `${story.title.slice(0, maxTitleLen - 3)}...`
				: story.title;
		lines.push(`  ${icon}  ${story.id}  ${storyTitle}`);
	}
	const completed = prd.userStories.filter((s) => s.passes).length;
	lines.push(dim(`  ${completed}/${prd.userStories.length} complete`));
	return lines.join("\n");
}

export async function statusCommand(prdFile: string, reviewDefault: number): Promise<void> {
	const prd = await readPrd(prdFile);
	console.log(render(prd, reviewDefault));
}

export async function confirm(
	prd: PRD,
	reviewDefault: number,
	tool: string,
	maxIter: number,
): Promise<boolean> {
	console.log(render(prd, reviewDefault));
	console.log();
	console.log(`Tool: ${tool} | Max iterations: ${maxIter}\n`);

	if (!process.stdin.isTTY) {
		return true;
	}

	process.stdout.write("Press Enter to start, or q + Enter to quit: ");
	for await (const chunk of console) {
		const line = chunk.toString().trim();
		return line !== "q" && line !== "Q";
	}
	return true;
}
